File: ext/examples/save_bruner.rs
//! This saves a resolution to Bruner's format. This saves the resulting files to the current
//! working directory. It is recommended that you run this in a dedicated subdirectory.

use algebra::module::Module;
use algebra::{Algebra, AlgebraType, MilnorAlgebra};
use ext::{chain_complex::ChainComplex, utils::query_module};
use fp::vector::prelude::*;
use itertools::Itertools;
use std::fmt::Write as _;
use std::fs::File;
use std::io::{BufWriter, Write as _};

fn main() -> anyhow::Result<()> {
    let resolution = query_module(Some(AlgebraType::Milnor), false)?;

    assert_eq!(*resolution.prime(), 2);
    let algebra = resolution.algebra();
    let algebra: &MilnorAlgebra = algebra.as_ref().try_into()?;

    let mut buffer = String::new();

    for s in 0..resolution.next_homological_degree() {
        let f = File::create(format!("hDiff.{s}"))?;
        let mut f = BufWriter::new(f);
        let module = resolution.module(s);
        // We don't use this when s = 0
        let dmodule = resolution.module(s.saturating_sub(1));
        let min_degree = module.min_degree();
        let max_degree = module.max_computed_degree();
        let num_gens: usize = (min_degree..=max_degree)
            .map(|t| module.number_of_gens_in_degree(t))
            .sum();

        writeln!(f, "        {num_gens}        {max_degree}\n")?;

        let d = resolution.differential(s);
        for t in min_degree..=max_degree {
            for idx in 0..module.number_of_gens_in_degree(t) {
                writeln!(f, "{t}\n")?;

                if s == 0 {
                    writeln!(f, "1\n0 0 1 i(0).\n\n\n")?;
                    continue;
                }
                let mut row_count = 0;
                buffer.clear();
                let dx = d.output(t, idx);

                let mut gen_count = 0;
                for gen_deg in min_degree..t {
                    for gen_idx in 0..dmodule.number_of_gens_in_degree(gen_deg) {
                        let op_deg = t - gen_deg;
                        let algebra_dim = algebra.dimension(op_deg);
                        let start = dmodule.generator_offset(t, gen_deg, gen_idx);
                        let slice = dx.slice(start, start + algebra_dim);
                        if slice.is_zero() {
                            gen_count += 1;
                            continue;
                        }
                        row_count += 1;
                        write!(buffer, "{gen_count} {op_deg} {algebra_dim} i").unwrap();
                        for (op_idx, _) in slice.iter_nonzero() {
                            let elt = algebra.basis_element_from_index(op_deg, op_idx);
                            write!(buffer, "({:?})", elt.p_part.iter().format(",")).unwrap();
                        }
                        writeln!(buffer, ".").unwrap();
                        gen_count += 1;
                    }
                }
                writeln!(f, "{row_count}")?;
                writeln!(f, "{buffer}")?; // buffer has one new line, writeln has one new line, add another one.
            }
        }
    }

    Ok(())
}

File: ext_py/examples/save_bruner.py
#!/usr/bin/env python3
"""Save a resolution to Bruner's format in the current working directory.

Python port of ext/examples/save_bruner.rs.
"""

import _query as query
from ext.algebra import AlgebraType
# NOTE: depends on MilnorAlgebra-typed resolution algebra and a SteenrodAlgebra ->
# MilnorAlgebra conversion (API_PROPOSAL §5.2); also on Resolution methods
# next_homological_degree / module / differential / prime and FreeModule
# {min_degree, max_computed_degree, number_of_gens_in_degree, generator_offset} and
# FreeModuleHomomorphism.output (API_PROPOSAL §7.1/§7.2/§5.3/§5.4).


def main():
    # standard backend: this example uses module(), unavailable on Nassau
    resolution = query.query_resolution(alg=AlgebraType.Milnor, algorithm="standard")
    resolution.compute_through_stem(query.query_n_s())

    assert resolution.prime == 2
    # Rust views the resolution's SteenrodAlgebra as a &MilnorAlgebra via try_into.
    # No such conversion is bound; we use the resolution's algebra directly, which
    # is the Milnor algebra and exposes dimension / basis_element_from_index.
    alg = resolution.algebra()

    for s in range(resolution.next_homological_degree()):
        with open(f"hDiff.{s}", "w") as f:
            module = resolution.module(s)
            # We don't use this when s = 0
            dmodule = resolution.module(max(s - 1, 0))
            min_degree = module.min_degree()
            max_degree = module.max_computed_degree()
            num_gens = sum(
                module.number_of_gens_in_degree(t)
                for t in range(min_degree, max_degree + 1)
            )

            f.write(f"        {num_gens}        {max_degree}\n\n")

            d = resolution.differential(s)
            for t in range(min_degree, max_degree + 1):
                for idx in range(module.number_of_gens_in_degree(t)):
                    f.write(f"{t}\n\n")

                    if s == 0:
                        f.write("1\n0 0 1 i(0).\n\n\n\n")
                        continue
                    row_count = 0
                    buffer = ""
                    dx = d.output(t, idx)

                    gen_count = 0
                    for gen_deg in range(min_degree, t):
                        for gen_idx in range(dmodule.number_of_gens_in_degree(gen_deg)):
                            op_deg = t - gen_deg
                            algebra_dim = alg.dimension(op_deg)
                            start = dmodule.generator_offset(t, gen_deg, gen_idx)
                            slice = dx.slice(start, start + algebra_dim)
                            if slice.is_zero():
                                gen_count += 1
                                continue
                            row_count += 1
                            buffer += f"{gen_count} {op_deg} {algebra_dim} i"
                            for op_idx, _ in slice.iter_nonzero():
                                elt = alg.basis_element_from_index(op_deg, op_idx)
                                buffer += "(" + ",".join(str(x) for x in elt.p_part) + ")"
                            buffer += ".\n"
                            gen_count += 1
                    f.write(f"{row_count}\n")
                    # buffer has one new line, writeln has one new line, add another one.
                    f.write(f"{buffer}\n")


if __name__ == "__main__":
    main()
